### PR TITLE
[Admin] Add files admin

### PR DESCRIPTION
**Pull request by illusionOfParadise**

### CODE REVIEW

File: src/Admin/Filesystem/FileAdmin.php [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+                        return 'file.type.'.$choice;
+                    },
+                ],
+            ])
+        ;
+    }
+
+    protected function configureFormFields(FormMapper $formMapper)
+    {
+        $isCreation = null === $this->getSubject() || null === $this->getSubject()->getId();
```
Comment (ottaviano) [inline src/Admin/Filesystem/FileAdmin.php:171]: could you use `App\Admin\AbstractAdmin::isCreation()` method?

File: src/Admin/Filesystem/FileAdmin.php [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+        ;
+    }
+
+    public function submit(FormEvent $event): void
+    {
+        /** @var File $file */
+        $file = $event->getData();
+        $uploadedFile = $file->getFile();
+
+        $this->fileManager->update($file, $uploadedFile);
```
Comment (ottaviano) [inline src/Admin/Filesystem/FileAdmin.php:249]: if an error of form validation appears we will keep the file on storage? 
what about saving the updated file on `postPersist` instead?

File: src/Entity/Filesystem/File.php [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+use Gedmo\Sluggable\Util\Urlizer;
+use Gedmo\Timestampable\Traits\TimestampableEntity;
+use Ramsey\Uuid\Uuid;
+use Symfony\Component\HttpFoundation\File\UploadedFile;
+use Symfony\Component\Serializer\Annotation as SymfonySerializer;
+use Symfony\Component\Validator\Constraints as Assert;
+
+/**
+ * @ORM\Entity(repositoryClass="App\Repository\Filesystem\FileRepository")
+ * @ORM\Table(indexes={
```
Comment (ottaviano) [inline src/Entity/Filesystem/File.php:20]: what about prefixing new tables by `filesystem_` too?

File: src/Entity/Filesystem/File.php [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+     * @ORM\JoinColumn(onDelete="SET NULL")
+     */
+    private $updatedBy;
+
+    /**
+     * @var bool
+     *
+     * @ORM\Column(type="boolean", options={"default": true})
+     */
+    private $displayed = true;
```
Comment (ottaviano) [inline src/Entity/Filesystem/File.php:107]: should we check this value on download page?

File: src/Entity/Filesystem/File.php [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+     * @Assert\Length(max=50)
+     */
+    private $mimeType;
+
+    /**
+     * @var int|null
+     *
+     * @ORM\Column(type="integer", options={"unsigned": true}, nullable=true)
+     *
+     * @Assert\LessThan(
```
Comment (ottaviano) [inline src/Entity/Filesystem/File.php:158]: IMO as it is not a data provided by the user we can avoid to validate it (and you are doing it with `File` assert)

File: src/Entity/Filesystem/File.php [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+    public function addPermission(FilePermission $permission): void
+    {
+        if (!$this->permissions->contains($permission)) {
+            $permission->setFile($this);
+            $this->permissions->add($permission);
+        }
+    }
+
+    public function removePermission(FilePermission $permission): void
+    {
```
Comment (ottaviano) [inline src/Entity/Filesystem/File.php:267]: do you need to call `setFile(null)` on $permission?

File: src/Entity/Filesystem/File.php [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+            case FileTypeEnum::EXTERNAL_LINK:
+                return (string) $this->parent ? $this->parent->getFullname().'/'.$this->getFullname() : $this->getFullname();
+            case FileTypeEnum::DIRECTORY:
+                return $this->getFullname();
+            default:
+                return null;
+        }
+    }
+
+    public function getFilenameForDownload(): ?string
```
Comment (ottaviano) [inline src/Entity/Filesystem/File.php:353]: what about moving this into controller?

File: src/Entity/Filesystem/FilePermission.php [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+     */
+    private $file;
+
+    /**
+     * @var string
+     *
+     * @ORM\Column(length=50)
+     *
+     * @Assert\NotBlank
+     * @Assert\Length(max=50)
```
Comment (ottaviano) [inline src/Entity/Filesystem/FilePermission.php:39]: do we really check the length if we check `Assert\Choice`?

File: src/Filesystem/FileManager.php [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        if ($file->getExternalLink()) {
+            $file->setType(FileTypeEnum::EXTERNAL_LINK);
+            $file->setOriginalFilename(null);
+            $file->setExtension(null);
+            $file->setMimeType(null);
+            $file->setSize(null);
+        }
+
+        if ($uploadedFile instanceof UploadedFile) {
```
Comment (ottaviano) [inline src/Filesystem/FileManager.php:38]: we can use `$file->getFile()` here, no?

File: src/Filesystem/FileManager.php [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        if (!$file->getFile() instanceof UploadedFile) {
+            throw new \RuntimeException(sprintf('The file must be an instance of %s', UploadedFile::class));
+        }
+
+        $this->storage->put($file->getPath(), file_get_contents($file->getFile()->getPathname()));
+    }
+
+    public function remove(File $file): void
+    {
+        $children = $this->entityManager->getRepository(File::class)->findBy(['parent' => $file]);
```
Comment (ottaviano) [inline src/Filesystem/FileManager.php:58]: could you use `$this->repository->`

File: src/Admin/Filesystem/FileAdmin.php [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+    public function prePersist($file)
+    {
+        parent::prePersist($file);
+
+        $this->fileManager->update($file);
+    }
+
+    public function postPersist($file): void
+    {
+        if ($file instanceof File && $file->getFile() instanceof UploadedFile) {
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Admin/Filesystem/FileAdmin.php:257]: 2/ why do we call `$this->fileManager->upload($file);` only for updated files and not for links? maybe remove the If like in `preUpdate`?

File: src/Admin/Filesystem/FileAdmin.php [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+    public function preUpdate($file)
+    {
+        parent::preUpdate($file);
+
+        $this->fileManager->update($file);
+    }
+
+    public function postUpdate($file): void
+    {
+        if ($file instanceof File) {
```
Comment (ottaviano) [inline src/Admin/Filesystem/FileAdmin.php:271]: same here, IMO we can only have File instance here

File: src/Admin/Filesystem/FileAdmin.php [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+                $this->fileManager->removeFromStorage($file);
+            }
+        }
+    }
+
+    public function preRemove($object)
+    {
+        parent::preRemove($object);
+
+        $this->children = $this->repository->findBy(['parent' => $object]);
```
Comment (ottaviano) [inline src/Admin/Filesystem/FileAdmin.php:284]: do we need to check if $file is a directory before execute this query?

File: src/Controller/Admin/AdminFileController.php [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+use Symfony\Component\Routing\Annotation\Route;
+
+class AdminFileController extends Controller
+{
+    /**
+     * @Route("/file-directory/autocompletion",
+     *     name="app_autocomplete_file_directory",
+     *     condition="request.isXmlHttpRequest()",
+     *     methods={"GET"}
+     * )
```
Comment (ottaviano) [inline src/Controller/Admin/AdminFileController.php:19]: `@Security(ADMIN...._FILE_...)` is missing

File: src/Controller/Admin/AdminFileController.php [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+     */
+    public function directoriesAutocompleteAction(Request $request, FileRepository $repository): JsonResponse
+    {
+        $zones = $repository->findForAutocomplete(
```
Comment (ottaviano) [inline src/Controller/Admin/AdminFileController.php:23]: $zones => $files

File: src/Entity/Filesystem/File.php [hunk 16]
```diff
@@ -0,0 +1,397 @@
...
+     * @ORM\Column(length=10, nullable=true)
+     *
+     * @Assert\Length(max=10)
+     */
+    private $extension;
+
+    /**
+     * @var string|null
+     *
+     * @ORM\Column(length=50, nullable=true)
```
Comment (ottaviano) [inline src/Entity/Filesystem/File.php:154]: `application/vnd.openxmlformats-officedocument.spreadsheetml.sheet` takes 66 chars, IMO we can remove these limites

File: src/Entity/Filesystem/File.php [hunk 17]
```diff
@@ -0,0 +1,397 @@
...
+use Symfony\Component\Validator\Constraints as Assert;
+
+/**
+ * @ORM\Entity(repositoryClass="App\Repository\Filesystem\FileRepository")
+ * @ORM\Table(name="filesystem_file", indexes={
+ *     @ORM\Index(columns={"type"}),
+ *     @ORM\Index(columns={"name"})
+ * })
+ *
+ * @AssertUniqueDirectory
```
Comment (ottaviano) [inline src/Entity/Filesystem/File.php:25]: what about using `@UniqueEntity(fields={"name"}, repositoryMethod="findDirectoryByName")`?

File: src/Admin/Filesystem/FileAdmin.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (illusionOfParadise) [reply src/Admin/Filesystem/FileAdmin.php:257]: It works not only for files, but for links also.
But you are right, there is no need of this check. I forgot to remove it after the refacto.

[the remark this answers is not in the document]
Comment (illusionOfParadise) [reply src/Admin/Filesystem/FileAdmin.php:257]: For `postPersist` it's good to call `upload` only in the case when `getFile` contains an `UploadedFile`.

[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Admin/Filesystem/FileAdmin.php:257]: oh sorry, I misunderstand between 2 methods `upload` & `update` 👍

File: src/Entity/Filesystem/File.php (hunk 16 above)
Comment (illusionOfParadise) [reply src/Entity/Filesystem/File.php:154]: Nice catch!
It's better to increase the column size.

**Comment by illusionOfParadise** [comment]
[ch12628]

**Review by ottaviano — APPROVED** [review]

**Pull request merged**